pkg/ipmi/ipmi.go

```diff
@@ -173,3 +195,96 @@ func (i *IPMI) ShutoffWatchdog() error {
 
 	return nil
 }
+
+func (i *IPMI) waitForSetInProgressCleared() error {
+	req := &req{}
+	req.msg.cmd = _GET_SYSTEM_INFO_PARAMETERS
+	req.msg.netfn = _IPMI_NETFN_APP
+	var data [4]byte
+
+	data[0] = 0
+	data[1] = _SET_IN_PROGRESS
+	data[2] = 0
+	data[3] = 0
+	req.msg.data = unsafe.Pointer(&data)
+	req.msg.dataLen = 4
+
+	retry := 20
+	for retry > 0 {
+		recv, err := i.sendrecv(req)
+		if err != nil {
+			return err
+		}
+		// response data byte 3 bit[1:0] == 00b indicates set complete
+		if len(recv) == 3 && (recv[2]&0x3) == 0 {
+			return nil
+		}
+		time.Sleep(1 * time.Millisecond)
+		retry--
+	}
+
+	return fmt.Errorf("Wait for Set in Progress cleared timeout")
+
+}
+
+func (i *IPMI) SetSystemFWVerToBmc() error {
+	strlen_max := 64 // Set 64 Bytes as the maximal version string length
+	entryData, tableData, err := getSmbiosData()
+	if err != nil {
+		return err
+	}
+
+	si, err := smbios.ParseInfo(entryData, tableData)
+	if err != nil {
+		return fmt.Errorf("error parsing data: %v", err)
+	}
+	t0, err := si.GetBIOSInformation()
+	if err != nil {
+		return fmt.Errorf("error Geting BIOS Information: %v", err)
+	}
+
+	fmt.Printf("Firmware Version: %v\n", t0.Version)
+	len := len(t0.Version)
+	if len > strlen_max || len == 0 {
+		fmt.Printf("Version length is 0 or longer than the suggested maximal length %d\n", strlen_max)
+		return nil
+	}
+
+	req := &req{}
+	req.msg.cmd = _SET_SYSTEM_INFO_PARAMETERS
+	req.msg.netfn = _IPMI_NETFN_APP
+	var data setSystemInfoReq
+	var copied int
+	data.paramSelector = _SYSTEM_FW_VERSION
+	data.setSelector = 0
+	index := 0
+	for len > 0 {
```

Comment:
I think it would be better to slit this and have a function doing the write similar to:
https://github.com/ipmitool/ipmitool/blob/master/lib/ipmi_mc.c#L1502

And maybe another one handling the split similar to:
https://github.com/ipmitool/ipmitool/blob/master/lib/ipmi_mc.c#L1566

I say maybe another one handling the split because it is not really needed in our case right now because we can only write 14bytes on the platforms we have.

Let me know if it makes sense.

Reply:
I think the code's structure is already very similar to it, the function in
https://github.com/ipmitool/ipmitool/blob/master/lib/ipmi_mc.c#L1502
simply passes the parameters and buffer pointer to sendrecv() which still need to be setup by the calling function ipmi_sysinfo_main().

Reply:
I think this can be better structured. What if somebody wants to set the `primary_os_name` instead of the `system_fw_version`. This is not modular enough and I think that splitting it will make it more readable.

How about having a `setsysinfo` function that takes a `setSystemInfoReq` struct as parameter?

I think it will make it more extensible and readable.

Reply:
So are you suggesting that:
Changing SetSystemFWVersion(version string) to a more generic function 
SetSysinfo(req setSystemInfoReq) and let the calling function decide what to set through the argument, as for the function body it still runs the for-loop to send, correct?

Reply:
Not moving the for-loop to that function.

In my opinion this should be split in two functions:

* `func (i *IPMI) setsysinfo(data *setSystemInfoReq) error`

To handle 1 write, not tied to setting the `system_fw_version`.

* `func (i *IPMI) SetSystemFWVersion(version string) error`

Which will have the for-loop / string splitting.

Let me know if it makes sense.

Reply:
Added a commit for splitting the function, please check.